### PR TITLE
Exception 2.0 prototype

### PR DESCRIPTION
**Pull request by lmazuel**

### CODE REVIEW

File: sdk/core/azure-core/azure/core/exceptions.py [hunk 1]
```diff
@@ -82,13 +83,96 @@ def map_error(status_code, response, error_map):
...
+
+        # Required fields, but assume they could be missing still to be robust
+        self.code = json_object.get(cls.CODE_LABEL)  # type: Optional[str]
+        self.message = json_object.get(cls.MESSAGE_LABEL)  # type: Optional[str]
+
+        if not (self.code or self.message):
+            raise ValueError("Impossible to extract code/message from received JSON:\n"+json.dumps(json_object))
+
+        # Optional fields
+        self.target = json_object.get(cls.TARGET_LABEL, None)  # type: Optional[str]
```
Comment (johanste) [inline sdk/core/azure-core/azure/core/exceptions.py]: Nitpick: is there a reason that you have an explicit `None` default value here?

File: sdk/core/azure-core/azure/core/exceptions.py [hunk 2]
```diff
@@ -82,13 +83,96 @@ def map_error(status_code, response, error_map):
...
+    def __str__(self):
+        return "({}) {}".format(self.code, self.message)
+
+    def message_details(self):
+        """Return a detailled string of the error.
+        """
+        # () -> str
+        error_str = "Code: {}".format(self.code)
+        error_str += "\nMessage: {}".format(self.message)
+        if self.target:
```
Comment (johanste) [inline sdk/core/azure-core/azure/core/exceptions.py]: Nitpick: technically, `self.code` and `self.message` are both optional, yet we don't guard those/print 'None' as the value. But we suppress `self.details`. I assume that this is intentional.

File: sdk/core/azure-core/azure/core/exceptions.py (hunk 1 above)
Comment (lmazuel) [reply sdk/core/azure-core/azure/core/exceptions.py]: nope, let's save some bytes

File: sdk/core/azure-core/azure/core/exceptions.py (hunk 2 above)
Comment (lmazuel) [reply sdk/core/azure-core/azure/core/exceptions.py]: If you look at the `__init__`, I actually accept one of both code or message to be None, but not both at the same time. So the possible print for `__str__` are:
```
(code) message
(None) message
(code) None
```

IOW, they are supposed to be required, but I don't want to fail if they are not here, but I don't want to also have an empty ODataV4 (is `{"foo": "bla"}` a correct OdataV4, with just missing code message? :))

TLDR: on purpose

File: sdk/core/azure-core/azure/core/exceptions.py (hunk 1 above)
Comment (lmazuel) [reply sdk/core/azure-core/azure/core/exceptions.py]: fixed
